File: Graphs/DFS/DFS.java
import java.util.*;


public class DFS {

    public static void dfs(node  start){
        ArrayDeque<node> s = new ArrayDeque<node>();

        s.push(start);
        while(s.isEmpty() == false){
            node top = s.pop();
            
            if(top.visited == false){
                top.visited = true;
                System.out.println("Visit " + top.name);
            }
            ArrayList<node> n;
            n = top.neighbors;
            for (node a: n){
                s.push(a);
            }

        }
    }
    
    public static void main(String args[]){
        node a = new node("A");
        node b = new node("B");
        node c = new node("C");
        node d = new node("D");

        ArrayList<node> la = new ArrayList<node>();
        la.add(b);
        la.add(c);
        ArrayList<node> lc = new ArrayList<node>();
        lc.add(d);

        a.neighbors = la;
        c.neighbors = lc;
        
        dfs(a);
    }

}

class node {
   ArrayList<node> neighbors;
   String name;
   boolean visited ;
   public node(String name){
        this.name = name; 
        this.visited = false;
        this.neighbors = new ArrayList<node>();
   }
};

File: Graphs/DFS/dfs.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>

using namespace std;

typedef struct node {
    string val;
    bool visited = false;
    vector < node >  neighbors ;
} node;


typedef vector < node >  list_nodes;



void dfs(node start){
    stack<node> s;
    s.push(start);
    while(s.empty() == false){
        node top = s.top();
        s.pop(); 

        if (top.visited == false){
            top.visited  = true;
            cout << "Node " << top.val << " visited" << endl;
            //Check for the termination condition
        }

        list_nodes n = top.neighbors;      
        for (node a: n){
            s.push(a);
        }


    }
}


int main(){

     /*          a
     *        /    \
     *       b      c
     *       |    /   \
     *       d   e     f
     *
     */

 
    node a,b,c,d,e,f;
    a.val = "a";
    b.val = "b";
    c.val = "c";
    d.val = "d";
    e.val = "e";
    f.val = "f";

    list_nodes la(2), lc(2), lb(1);

    
    lc[0] = e;
    lc[1] = f;
    c.neighbors = lc;

    lb[0] = d;
    b.neighbors = lb;


    la[0] = b;
    la[1] = c;    
    a.neighbors = la;
    


    cout<< " dfs -> a" <<endl<<endl;
    dfs(a);
    
    return 0; 

}
